Add option to control Biharmonic coeff. via grid Reynolds number

gustavo-marques opened the pull request:
This PR adds the option to specify the biharmonic horizontal viscosity coefficient (`AH`)
via a user-specified (Biharmonic) grid Reynolds number (input parameter `RE_AH`). 
If `RE_AH` is nonzero, `AH` is scaled so that the biharmonic Reynolds number is equal to this.
`RE_AH` is defined as:

`RE_AH = (U dx3)/AH,    (1)
`
where `dx3` is the harmonic mean of the squares of the grid^(3/2) [L3].

Following @s-ragen and @StephenGriffies suggestion, this PR also adds the option to diagnose both the Laplacian and Biharamonic grid Reynolds numbers. Eq (1) is used for the Biharamonic while the following is used for Laplacian:

`grid_Re_Kh = (U sqtr(dx2))/Kh, (2)
`
where `dx2` is the harmonic mean of the squares of the grid [L2].

Both grid_Re_Kh and grid_Re_Ah are 3D arrays. Examples of surface values for an ocean_only case with CESM/MOM6 are shown below:

![image](https://user-images.githubusercontent.com/11339137/79497455-f6510f80-7fe4-11ea-9041-0e804a9e25c6.png)


![image](https://user-images.githubusercontent.com/11339137/79497176-7d51b800-7fe4-11ea-8eeb-adfc0ffab431.png)

@alperaltuntas, I have not modified the OMP calls so we might have to do that later. This PR should not change answers since, by default, we are not using `RE_AH`.

gustavo-marques committed:
Add option to scale AH via a biharmonic Reynolds #

This is done via parameter RE_AH: if nonzero, the biharmonic
coefficient is scaled so that the biharmonic Reynolds number
is equal to this.

gustavo-marques committed:
Add diags for Lapl. and Bihar grid Reynolds #s

grid_Re_Kh = (U sqtr(dx2))/Kh

grid_Re_Kh = (U dx3)/Ah

where dx2 is the harmonic mean of the squares of the grid [L2], and
dx3 is the harmonic mean of the squares of the grid^(3/2) [L3]

alperaltuntas commented:
Looks like there is a floating point exception in symmetric mode: 

 tc2.a.symmetric: Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation

gustavo-marques commented:
I saw that, thanks. I will try to look at this later today.

gustavo-marques committed:
Fix indices in KE calculation

Hoping this will fix the floating point exception in symmetric mode.

gustavo-marques committed:
Avoid division by zero

This PR add two hard-coded parameters (AH_min and KH_min) to avoid
dividing by zero when computing the Biharmonic and Laplacian grid
Reynolds numbers, respectively.

It also fixed the size of an array used in the Biharmonic Re
calculation.

gustavo-marques committed:
Fix openmp directives

gustavo-marques committed:
Fix openmp directives

gustavo-marques committed:
Fix OMP calls

gustavo-marques commented:
@alperaltuntas, this PR is failing because of differences in chksum_diag (two new diagnostics have been added). I've confirmed with @adcroft that we should go ahead and do the merge. We will discuss how to handle this in the future on the next dev call. Thanks.

alperaltuntas reviewed (approved):
test: aux_mom/cheyenne/intel
